packages/core/src/context/context.ts

```diff
@@ -59,45 +62,79 @@ export const registerAll = (boundDependencies: BoundDependency<any, any>[]) => (
     context,
   );
 
+/**
+ * Resolves eager dependencies within the context
+ * @since v2.0.0
+ */
 export const resolve = async (context: Context): Promise<Context> => {
+  const resolveEagerDependency = ({ token, dependency }: BoundDependency<unknown, ContextEagerReader>): T.Task<Context> => pipe(
+    () => pipe(context, dependency.eval, d => Promise.resolve(d)),
+    T.chain(resolvedDependency => T.fromIO(() => context.set(token, resolvedDependency))));
+
   for (const [token, dependency] of context) {
-    context.set(
-      token,
-      isEagerDependency(dependency)
-        ? await Promise.resolve(dependency.eval(context))
-        : dependency,
-    );
+    if (!isEagerDependency(dependency)) continue;
+    await resolveEagerDependency({ token, dependency })();
   }
 
   return context;
 }
 
-export const lookup = (context: Context) => <T>(token: ContextToken<T>): O.Option<T> => pipe(
-  M.lookup(ordContextToken)(token, context),
-  O.map(dependency => {
-    if (!dependency.eval) {
-      return dependency;
-    }
-
-    const boostrapedDependency = isLazyDependency(dependency)
-      ? dependency.eval()(context)
-      : dependency.eval(context);
-
-    context.set(token, boostrapedDependency);
-    return boostrapedDependency;
-  }),
-);
+/**
+ * Lookup the dependency for a token in a `Context`
+ * @param context
+ * @param token
```

Comment:
nit-pick: this doc would imply a signature like `(Context, ContextToken<T>) -> Option<T>`, but its `Context -> ContextToken<T> -> Option<T>`.

Reply:
Same is true for the other functions :thinking:

Reply:
Yes, I'm aware of it. Maybe we should use Hindley-Milner like type annotations? But in this case there is no syntax support 🤨

Reply:
True, (though there are workarounds: jsdoc/jsdoc#1286) I would just leave it off, because this signature is just plain wrong and may lead to confusion. Its not necessary anyways in the context of TypeScript.